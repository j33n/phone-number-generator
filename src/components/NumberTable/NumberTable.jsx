import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { formZeroedNumber } from '../../pages/GenerateNumbers/GenerateNumbers';
import Loader from '../Loader/Loader';

import './NumberTable.scss';

const sortBy = {
  ASC: 'ASC',
  DESC: 'DESC',
  MIN: 'MIN',
  MAX: 'MAX',
};

class NumberTable extends PureComponent {
  static propTypes = {
    numbers: PropTypes.arrayOf(PropTypes.string).isRequired,
    sortable: PropTypes.bool,
  }

  static defaultProps = {
    sortable: false,
  }

  state = {
    numbers: [],
    sort: '',
    showLoader: false,
  }

  componentDidMount = () => {
    const { numbers } = this.props;
    this.setState({ numbers });
  }

  sortNumbersBy = (type) => {
    const { numbers } = this.props;
    const asc = numbers.sort();
    switch (type) {
      case 'ASC':
        return this.setState({
          numbers: asc,
        });
      case 'DESC':
        return this.setState({
          numbers: numbers.sort().reverse(),
        });
      default:
        return this.setState({ numbers });
    }
  }

  handleOnSort = (sort) => {
    this.setState({
      sort,
      showLoader: true,
    });
    setTimeout(() => {
      this.setState({ showLoader: false });
    }, 1000);
  }

  render() {
    const { sortable, } = this.props;
    const { numbers, sort, showLoader, } = this.state;
    const {
      DESC, ASC, MIN, MAX,
    } = sortBy;
    const toTabulize = () => {
      if (sort === ASC) {
        return numbers.sort();
      } if (sort === DESC) {
        return numbers.sort().reverse();
      } if (sort === MAX) {
        const maxNumber = Math.max(...numbers);
        return [formZeroedNumber(maxNumber, 10 - maxNumber.length)];
      } if (sort === MIN) {
        const minNumber = formZeroedNumber(Math.min(...numbers));
        return [formZeroedNumber(minNumber, 10 - minNumber.length)];
      }
      return numbers;
    };
    return (
      <div className="number-display">
        {showLoader && <Loader />}
        <div className="number-display--header">
          <h2>Here are all the numbers generated</h2>
        </div>
        {sortable
          && (
            <div className="sortable">
              <div role="presentation" onClick={() => this.handleOnSort(DESC)}>Desc</div>
              <div role="presentation" onClick={() => this.handleOnSort(ASC)}>Asc</div>
              <div role="presentation" onClick={() => this.handleOnSort(MIN)}>Min</div>
              <div role="presentation" onClick={() => this.handleOnSort(MAX)}>Max</div>
            </div>
          )
        }
        <div className="sub-heading">
          <div className="index-header">
            Index
          </div>
          <div className="number-header">
            Phone Number
          </div>
        </div>
        {!showLoader && toTabulize().map((number, index) => (
          <div className={`number-display--number number-display--number__${index % 2}`} key={number}>
            <div className="index-column">
              {index}
            </div>
            <div className="number-column">
              {number}
            </div>
          </div>
        ))}
      </div>
    );
  }
}

export default NumberTable;
