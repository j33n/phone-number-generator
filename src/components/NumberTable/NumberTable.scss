.number-display {
    display: flex;
    flex-direction: column;
    width: 70%;
    margin: auto;
    border-bottom: 1px solid rgb(210, 210, 210);

    &--header {
        border: 1px solid rgb(210, 210, 210);
        text-align: center;
        margin-top: 10px;

        h2 {
            margin: 5px auto;
        }
    }

    &--number {
        display: flex;
        flex-direction: row;
        text-align: center;

        &__0 {
            background: rgba(0, 0, 0, 0.04);
        }

        .index-column {
            width: 15%;
            padding: 5px 0;
            border: solid rgb(210, 210, 210);
            border-width: 0px 1px 0px 1px;
        }

        .number-column {
            width: 85%;
            padding: 5px 0;
            border: solid rgb(210, 210, 210);
            border-width: 0px 1px 0 0;
        }
    }

    .sub-heading {
        display: flex;
        width: 100%;
        text-align: center;

        .index-header {
            width: 15%;
            padding: 5px 0;
            border: solid rgb(210, 210, 210);
            border-width: 0px 1px 0px 1px;
        }

        .number-header {
            width: 85%;
            padding: 5px 0;
            border: solid rgb(210, 210, 210);
            border-width: 0px 1px 0 0;
        }
    }
}

.sortable {
    display: flex;
    width: calc(100% - 1px);
    text-align: center;
    border-bottom: 1px solid rgb(210, 210, 210);
    border-left: 1px solid rgb(210, 210, 210);
    div {
        width: calc(25% - 1px);
        border-right: 1px solid rgb(210, 210, 210);
        background: #eee;
        &:hover{
            background: #fff;
            cursor: pointer;
        }
    }
}
