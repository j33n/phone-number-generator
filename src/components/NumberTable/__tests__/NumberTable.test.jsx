import React from 'react';
import { shallow } from 'enzyme';

import NumberTable from '../NumberTable';

describe('NumberTable', () => {
  const numbers = ['1', '2', '3'];
  const component = <NumberTable numbers={numbers} sortable />;
  const tree = shallow(component);

  jest.useFakeTimers();

  it('renders', () => {
    expect(component).toMatchSnapshot();
  });

  it('should sort by `DESC`', () => {
    tree.find('#desc').simulate('click');
    expect(tree.state('sort')).toEqual('DESC');
    expect(tree.state('showLoader')).toBeTruthy();
    jest.advanceTimersByTime(1000);
    expect(tree.state('showLoader')).toBeFalsy();
    expect(tree.instance().toTabulize()).toEqual(['3', '2', '1']);
  });

  it('should sort by `ASC`', () => {
    tree.find('#asc').simulate('click');
    expect(tree.state('sort')).toEqual('ASC');
    expect(tree.state('showLoader')).toBeTruthy();
    jest.advanceTimersByTime(1000);
    expect(tree.state('showLoader')).toBeFalsy();
    expect(tree.instance().toTabulize()).toEqual(['1', '2', '3']);
  });

  it('should sort by `MIN`', () => {
    tree.find('#min').simulate('click');
    expect(tree.state('sort')).toEqual('MIN');
    expect(tree.state('showLoader')).toBeTruthy();
    jest.advanceTimersByTime(1000);
    expect(tree.state('showLoader')).toBeFalsy();
    expect(tree.instance().toTabulize()).toEqual(['0000000001']);
  });

  it('should sort by `MAX`', () => {
    tree.find('#max').simulate('click');
    expect(tree.state('sort')).toEqual('MAX');
    expect(tree.state('showLoader')).toBeTruthy();
    jest.advanceTimersByTime(1000);
    expect(tree.state('showLoader')).toBeFalsy();
    expect(tree.instance().toTabulize()).toEqual(['3']);
  });
});
