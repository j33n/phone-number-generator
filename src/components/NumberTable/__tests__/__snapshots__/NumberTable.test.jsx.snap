// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberTable renders 1`] = `
<NumberTable
  numbers={
    Array [
      "1",
      "2",
      "3",
    ]
  }
  sortable={true}
/>
`;
